File: LeetCode/permutations.cpp
class Solution {
public:
    vector< vector<int> > res;
    
    void solve(vector<int>& nums, int start) {
        if(start == nums.size()) {
            res.push_back(nums);
            return;
        }
        
        for(int i = start; i < nums.size(); i++) {
            swap(nums[start], nums[i]);
            solve(nums, start+1);
            swap(nums[start], nums[i]);
        }
    }
    
    vector<vector<int>> permute(vector<int>& nums) {
        solve(nums, 0);
        
        return res;
    }
};

File: LeetCode/permutations.go
func solve(nums []int, k int, ans *[][]int) {
    if k == len(nums) {
        tmp := make([]int, len(nums))
        copy(tmp, nums)
        *ans = append(*ans, tmp)
    }

    for i := k; i < len(nums); i++ {
        nums[i], nums[k] = nums[k], nums[i]
        solve(nums, k+1, ans)
        nums[i], nums[k] = nums[k], nums[i]
    }
}

func permute(nums []int) [][]int {
    ans := [][]int{}

    solve(nums, 0, &ans)

    return ans
}
